### PR TITLE
Added Group.tree method

### PR DESCRIPTION
**Pull request by TomAugspurger**
Adds a `Group.tree` method for pretty printing a hierarchy.

Screenshot 2024-10-24 at 8 05 59 AM

For now the old `expand` keyword, which I think controlled whether the tree was expanded by default in the HTML output, is not implemented.

We have one design choice to make: should `.tree` return an object with a custom `__repr__`, or should it return the string itself? My preference, implemented here, is returning the object. That makes it easier to pipe into other applications other than the Python REPL. The cost is a bit more expansion of the public API.

Closes https://github.com/zarr-developers/zarr-python/issues/2406

* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by TomAugspurger** [comment]
There was a `notebooks/repr_tree.ipynb` notebook that I just ran through. This PR does *not* implement

- A plain ASCII tree repr with `print(bytes(g1.tree()).decode())`. We kind of get that for free through rich, which will automatically detect whether the console being rendered to supports unicode or just ASCII. We won't have an API for doing that (we can probably document how to configure rich to get out an ascii repr if someone really wants that)
- An interactive HTML repr (I think for collapsing nodes). I've toyed around with the idea of a textual-based TUI for Zarr groups that would be a better spot for this kind of interactive explorer IMO.

File: src/zarr/core/_tree.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    ) -> dict[str, str]:
+        # For jupyter support.
+        # Unsure why mypy infers the return type to by Any
+        return self._tree._repr_mimebundle_(include=include, exclude=exclude, **kwargs)  # type: ignore[no-any-return]
+
+
+async def group_tree_async(group: AsyncGroup, max_depth: int | None = None) -> TreeRepr:
+    tree = rich.tree.Tree(label=f"[bold]{group.name}[/bold]")
+    nodes = {"": tree}
+    members = sorted([x async for x in group.members(max_depth=max_depth)])
```
Comment (jhamman) [inline src/zarr/core/_tree.py:46]: cc @d-v-b who is looking at making this faster 🚀

**Review by jhamman — APPROVED** [review]
Looks great @TomAugspurger!

File: src/zarr/core/_tree.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+except ImportError as e:
+    raise ImportError("'rich' is required for Group.tree") from e
+
+
+class TreeRepr:
+    """
+    A simple object with a tree-like repr for the Zarr Group.
+
+    Note that this object and it's implementation isn't considered part
+    of Zarr's public API.
```
Comment (jhamman) [inline src/zarr/core/_tree.py:20]: 👍

**Pull request merged**